Render live data on /dashboard

lpatmo opened the pull request:
On /dashboard:
- [x] cards look like:
![image](https://user-images.githubusercontent.com/4512699/55350395-db127100-5470-11e9-8904-2a299faaf875.png)
- [x] When a category on the sidebar is clicked on, the entries opposite of that category are displayed.
- [x] All the entries are live from the database
- [ ] In the "all" category, show only the first three recent entries under each section, and a "Show all" link
![image](https://user-images.githubusercontent.com/4512699/55350526-2dec2880-5471-11e9-9723-4ecfef76a675.png)

Commit:
Pass category ID number from state up into withTracker

Commit:
Pass down entries from withTracker to dashboard_cards_section

Commit:
Make sure to pass down 'all' entries if 'all' is selected

Commit:
Pass entries down into cards, and display default text if no entries found

Commit:
Fix section header names -- need to be the opposite of what the user selected

Commit:
Added _cards.scss style and added CardDeck

Commit:
Fix bug when 'all' is clicked on

Commit:
Show current unmathced cards created by user, and prevent button from appearing w/ a prop on that cards component

Commit:
Apply CSS fix suggestion from @angelocordon

Commit:
move timezone to the bottom of the card and change 'matched' to be a string

Commit:
comment out the console logs

Commit:
Remove limit on each category
Remove limit on category until pagination can be limited

Commit:
Rename 'Request Connection' to be 'Request Match'

Commit:
Rename 'past matches' to be 'Other Submissions'

angelocordon commented on imports/ui/components/board/board.jsx:
```diff
@@ -13,7 +13,7 @@ class Board extends React.Component {
   getColumns() {
     const { entries = [] } = this.props;
-
+    //console.log(this.props.entries)
```
Please make sure you remove all the `console.logs` when calling this PR done (not just commenting them out) :D

lpatmo replied:
Can we wait to remove them until after pagination is in? I still need them around.

angelocordon replied:
Yeah! So long we do it before it gets merged 👍

angelocordon commented on imports/ui/components/dashboard/dashboard_cards_section.jsx:
```diff
@@ -1,21 +1,29 @@
+            People looking for {sectionMapper[section]}
+          </Card.Header>  
+          <Card.Body as={CardColumns}>
+               {entries.length ? entries.map((entry, i) => (
```
I think we should prefer map functions as separate methods instead of evaluating them inside the `return`. I think they are much more readable this way: 

```jsx
render () {
   const { section, ... } = this.props
   const sectionMapper = { ... }

   const matchCards = () => {
      if (entries.length > 0) {
         return entries.map((entry, i) => {
             <MatchCard key={i} lookingFor ... />
         })
      } else {
         return <Card.Text>No active entries</Card.Text>
      }
   }

   return (
       ...
          <Card.Body as={CardColumns}>
              {matchCards}
          </Card.Body>
       ...
   )
}
```

angelocordon replied:
Additionally, AirBnB had a good post on not using indexes as keys: 
https://medium.com/@robinpokorny/index-as-a-key-is-an-anti-pattern-e0349aece318

lpatmo replied:
Thanks for the article! What do you think about using `entry._id` as the key?

lpatmo replied:
I'm not convinced that ```const matchCards = () => {
      if (entries.length > 0) {
         return entries.map((entry, i) => {
             <MatchCard key={i} lookingFor ... />
         })
      } else {
         return <Card.Text>No active entries</Card.Text>
      }
   }``` is a lot more readable than using a ternary operator, which seems to be standard in React, though.

angelocordon replied:
So long as `entry._id` is actually unique to that `entry` and not its index placement in the array 👍

angelocordon replied:
Re: ternary operators -- I'm not fully convinced it's that much of a standard either 😅 It just seems like bad practice and a lot of examples has just normalized it. Ternary operators are really more meant for a single line evaluation. 

From the [docs](https://reactjs.org/docs/conditional-rendering.html#inline-if-else-with-conditional-operator): 
> It can also be used for larger expressions although it is less obvious what’s going on

In any case, the main goal is to ensure that our code is readable and easy to understand regardless of how much React experience a contributor has. It would be much better to write out some proper if/else statements as it's clear exactly what's happening. 

(See example from: 
https://reactjs.org/docs/conditional-rendering.html#element-variables
https://codepen.io/gaearon/pen/QKzAgB?editors=0010)

angelocordon commented on imports/ui/components/dashboard/match_card.jsx:
```diff
@@ -1,33 +1,38 @@
-          <Card.Title className='font-weight-normal text-capitalize'>
-            { this.props.name }
+          <Card.Title className='font-weight-normal'>
+          { this.props.oneLineIntro }
```
I started deconstructing props at the beginning of a `render()` method as I thought it's made the JSX more readable. Any thoughts on establishing that as a pattern?

```jsx
render () {
   const { oneLineIntro, lookingFor } = this.props

   return (
      <Card>
          <Card.Body>
              <Card.Title className='font-weight-normal'> { oneLineIntro } </Card.Title>
              <Card.Text> { lookingFor } </Card.Text>
          ...
   )
}

lpatmo replied:
++ Thanks, I've made this change.

angelocordon commented on imports/ui/styles/_cards.scss:
```diff
@@ -0,0 +1,8 @@
+.card-deck .card {
```
Prefer to lean in on Bootstrap's utility methods instead of making new ones - https://getbootstrap.com/docs/4.0/utilities/spacing/

If you need the MatchCard component to have a margin bottom, try adding a class to the `<Card>` component. Try using `.mb-2` (it should go up to 5, I believe). 

Similarly, you shouldn't need to set a new `small` CSS rule, try using `<p><small> ... </small></p>` or possibly the `.small` class. 
https://getbootstrap.com/docs/4.0/content/typography/#inline-text-elements

lpatmo replied:
Nice, I didn't know about the Bootstrap's spacing method! I removed the .card-deck .card. The `<small>` I defined I was trying to override, since the current `small` rule isn't small enough. 

```<Card.Footer><small className="text-muted">{ timezone }</small></Card.Footer>```

![image](https://user-images.githubusercontent.com/4512699/55379377-c2867300-54d1-11e9-8d44-43a22be6f962.png)

angelocordon replied:
Hm, granted the font-weight and muted classing, I'm afraid it might actually be too small and too light for it to be accessible / readable:

![image](https://user-images.githubusercontent.com/3891869/55380716-e3e95e00-54d5-11e9-906d-17fb4235df37.png)

imo, it's not really worth overriding :\

angelocordon commented on imports/ui/components/dashboard/dashboard.jsx:
```diff
@@ -3,11 +3,14 @@ import DashboardSidebar from './dashboard_sidebar'
-  'oss-contributors': 'oss-projects'
+const sectionTargets = { //see: categories.js for numbers
+  'all': 'all',
+  'mentors' : 3,
```
I'm not quite sure I understood why we need to identify the sections with number IDs from checking out the categories.js file and this commit https://github.com/codebuddies/cb-connect/commit/ffb488e8e3b669cf29faf90b9524f6de2842bba7.

Did it have something to do with using them as IDs to pass to `withTracker()` for Mongo? 

If we really do have a good reason to use numerical values for this, let's at least add additional notes for our reasoning.

lpatmo replied:
They needed to be numbers because in our query to the database, we were looking for something like:

```
Entries.find({userId: {$ne: currentUser._id}, "category.id": "3"}).fetch()
```

(Replace "3" with another number to indicate another category.)

angelocordon replied:
Ah! Makes more sense now, thanks! Would be helpful to add that as a comment here. 

This is interesting -- it's a bit confusing here as in the, it looks like `category.id` is defined as a String based on the Entry schema:
https://github.com/codebuddies/cb-connect/blob/8611a9a4338a7795801f833250f97452f317c766/imports/api/entries/entries.js#L9-L10

Considering we're most likely not going to have _a lot_ of other categories, does it make more sense to just the category title as the id? (not important in this PR now but perhaps a separate discussion down the line).

lpatmo replied:
The problem with using `category.title` as the ID is that the `title` might change if we decide to rename any of the short labels in `lib/data/categories.js`. IMO it's safer to refer to the categories by their numbers, since for example one day we might want to rename `short_label: 'Contributor(s)',` to be `short_label: 'Open Source Contributors'. 

Along the same lines, in the most recent batch of commits, I removed references to a `section` array and imported  `lib/data/categories.js`, so that we weren't rewriting the section labels but referring back to the `lib/data/categories.js` source.

angelocordon replied:
I didn't actually know about `categories.js` until this branch - perhaps if we're trying to make sure that we can keep our categories always in synched, we should just import the categories from that lib file so that whenever we update it, it will always be updated everywhere else. Thoughts?

lpatmo replied:
I think that would be ideal. :) I did it in dashboard.jsx with ```    const sections = categories.map((section) => {
        return section.short_label;
    })``` but two other places I left because the refactoring would have taken a bit more work, and I was thinking we'd come back to it later. At the moment, if I update the short_label strings in categories.js, it will break the other other parts of the files that reference those string names directly, so part of the refactoring === changing the references to be the category numbers, instead of the strings.

angelocordon reviewed (commented):
Left some feedback, but otherwise, looking like it's going towards a good direction 🙌

Commit:
Deconstruct props in match_cards.js

Commit:
Use bootstrap helper class instead of defining .card-deck .card margin

Commit:
Use entry ID as key instead of counter index as key

Commit:
Remove feedback/career advice as a category

Commit:
Remove the unimportant console logs

Commit:
Fix issue with all entries displaying under each section when 'all' is clicked on
- Fix spelling of 'label'
- Use the imported categories.js when available
- Give 'Other' a category ID of 2
- Pass currentUserName down as a context variable

angelocordon commented on imports/ui/components/dashboard/matches_section.jsx:
```diff
@@ -13,19 +15,19 @@ const MatchesSection = () => {
             </Nav.Item>
             <Nav.Item>
-              <Nav.Link eventKey='pastMatches'>Past Matches</Nav.Link>
+              <Nav.Link eventKey='otherSubmissions'>Other Submissions</Nav.Link>
```
Maybe instead of `submissions`, we should keep our vocabulary consistent and call it `entries`?

lpatmo replied:
Sounds good!

angelocordon commented on imports/ui/containers/dashboard.jsx:
```diff
@@ -3,13 +3,24 @@ import { withTracker } from 'meteor/react-meteor-data';
+  const currentUser = Meteor.user();
+  const currentUserName = Meteor.user().profile.name;
+  const categoryIdString = categoryId.get().toString();
+  const entries = categoryIdString === 'all' ? Entries.find().fetch() : Entries.find({userId: {$ne: currentUser._id}, "category.id": categoryIdString}).fetch()
```
I'd implore to refactor this so that it fits within an 80 character line limit.

lpatmo replied:
Done! I broke this up into multiple lines.

angelocordon commented on imports/ui/styles/_cards.scss:
```diff
@@ -0,0 +1,5 @@
+small {
```
I think this is way too small and unreadable from an accessibility standpoint

![image](https://user-images.githubusercontent.com/3891869/55602247-87768080-5719-11e9-9c55-9bd7edc57aa9.png)

in comparison to using some built in bootstrap markup

![image](https://user-images.githubusercontent.com/3891869/55602432-66faf600-571a-11e9-930f-60671634e239.png)

```html
<div class="card-footer text-right">
    <small class="text-muted">(GMT-12:00) International Date Line West</small>
</div>
```

lpatmo replied:
What about 11px?

This was what I had in my HTML:
      ```  <Card.Footer><small className="text-muted">{ timezone }</small></Card.Footer>```

But the default caused the timezone to be on two lines:
![image](https://user-images.githubusercontent.com/4512699/55602941-fbfeee80-571c-11e9-9af7-df2fc3d6d297.png)

Maybe we can decrease the line height here?

lpatmo replied:
Put this into a separate issue: https://github.com/codebuddies/cb-connect/issues/48

angelocordon reviewed (changes requested):
A few small changes, but looks good 👍

Commit:
Prevent users' own cards from being displayed when 'all' is clicked on in the sidebar, and remove  CSS styling

Commit:
Break 'entries' line to be more readable
